File: app/src/main/java/ch/bailu/aat/activities/AbsBackButton.java
package ch.bailu.aat.activities;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;

public class AbsBackButton extends AbsActivity {

    @Override
    public void onBackPressed() {
        View view = getWindow().getDecorView();

        if (onBackPressed(view) == false)
            super.onBackPressed();

    }

    public void onBackPressedMenuBar() {
        super.onBackPressed();
    }

    public interface OnBackPressed {
        boolean onBackPressed();
    }

    public abstract static class OnBackPressedListener extends View implements OnBackPressed {
        public OnBackPressedListener(Context context) {
            super(context);
            setVisibility(INVISIBLE);
        }

        @Override
        public abstract boolean onBackPressed();
    }

    private boolean onBackPressed(View view) {
        if (view instanceof OnBackPressed) {
            if (((OnBackPressed) view).onBackPressed())
                return true;
        }

        if (view instanceof ViewGroup) {
            if (onBackPressedChildren((ViewGroup) view))
                return true;
        }
        return false;
    }

    private boolean onBackPressedChildren(ViewGroup parent) {
        int count = parent.getChildCount();

        for (int i=0; i < count; i++) {
            View view = parent.getChildAt(i);
            if (onBackPressed(view)) return true;
        }
        return false;
    }
}

File: aat-android/src/main/java/ch/bailu/aat/activities/AbsBackButton.kt
package ch.bailu.aat.activities

import android.content.Context
import android.view.View
import android.view.ViewGroup

abstract class AbsBackButton : AbsActivity() {
    override fun onBackPressed() {
        if (!onBackPressed(window.decorView)) super.onBackPressed()
    }

    fun onBackPressedMenuBar() {
        super.onBackPressed()
    }

    abstract class OnBackPressedListener(context: Context) : View(context){
        init {
            visibility = INVISIBLE
        }

        abstract fun onBackPressed(): Boolean
    }

    private fun onBackPressed(view: View): Boolean {
        if (view is OnBackPressedListener) {
            if (view.onBackPressed()) return true
        }
        return if (view is ViewGroup) {
            onBackPressedChildren(view)
        } else false
    }

    private fun onBackPressedChildren(parent: ViewGroup): Boolean {
        val count = parent.childCount
        for (i in 0 until count) {
            val view = parent.getChildAt(i)
            if (onBackPressed(view)) return true
        }
        return false
    }
}
